rudder-web/src/main/scala/com/normation/rudder/web/rest/compliance/ComplianceAPIService.scala

```diff
@@ -0,0 +1,265 @@
+* the GNU Affero GPL v3, the copyright holders add the following
+* Additional permissions:
+* Notwithstanding to the terms of section 5 (5. Conveying Modified Source
+* Versions) and 6 (6. Conveying Non-Source Forms.) of the GNU Affero GPL v3
+* licence, when you create a Related Module, this Related Module is
+* not considered as a part of the work and may be distributed under the
+* license agreement of your choice.
+* A "Related Module" means a set of sources files including their
+* documentation that, without modification of the Source Code, enables
+* supplementary functions or services in addition to those offered by
+* the Software.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU Affero General Public License for more details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* along with this program. If not, see <http://www.gnu.org/licenses/agpl.html>.
+*
+*************************************************************************************
+*/
+
+package com.normation.rudder.web.rest.compliance
+
+import com.normation.inventory.domain.NodeId
+import com.normation.rudder.domain.policies.RuleId
+import com.normation.rudder.domain.reports.ComplianceLevel
+import com.normation.rudder.repository.RoNodeGroupRepository
+import com.normation.rudder.repository.RoRuleRepository
+import com.normation.rudder.services.nodes.NodeInfoService
+import com.normation.rudder.services.reports.ReportingService
+import net.liftweb.common.Box
+import com.normation.rudder.domain.policies.Rule
+import com.normation.rudder.domain.nodes.NodeInfo
+import net.liftweb.common.Full
+
+
+
+/**
+ * The class in charge of getting and calculating
+ * compliance for all rules/nodes/directives.
+ */
+class ComplianceAPIService(
+    rulesRepo       : RoRuleRepository
+  , nodeInfoService : NodeInfoService
+  , nodeGroupRepo   : RoNodeGroupRepository
+  , reportingService: ReportingService
+) {
+
+
+
+  /**
+   * Get the compliance for everything
+   */
+ private[this] def getByRulesCompliance(rules: Set[Rule]) : Box[Seq[ByRuleRuleCompliance]] = {
+
+    for {
+      groupLib  <- nodeGroupRepo.getFullGroupLibrary()
+      nodeInfos <- nodeInfoService.getAll()
+    } yield {
+
+
+      // get an empty-initialized array of compliances to be used
+      // as defaults
+      val initializedCompliances : Map[RuleId, ByRuleRuleCompliance] = {
+        (rules.map { rule =>
+          val nodeIds = groupLib.getNodeIds(rule.targets, nodeInfos)
+
+          (rule.id, ByRuleRuleCompliance(
+              rule.id
+            , ComplianceLevel(noAnswer = nodeIds.size)
+            , Seq()
+          ))
+        }).toMap
+      }
+
+      //reports by rules
+      //that's a *little* gruikkk because, well,
+      //we just plainly ignore errors
```

Comment:
hmmm What do you think of...

    If there was an error (so partition not empty)
      logger.error "there was error when   computing compliance for rules x, y z, please enable debug output for more details"
      for each error 
        logger.debug(" blah blah error")